service/storage_service.cc

```diff
@@ -5358,6 +5359,100 @@ future<> storage_service::update_fence_version(token_metadata::version_t new_ver
     });
 }
 
+static
+locator::tablet_replica get_leaving_replica(const locator::tablet_info& tinfo, const locator::tablet_transition_info& trinfo) {
+    std::unordered_set<locator::tablet_replica> leaving(tinfo.replicas.begin(), tinfo.replicas.end());
+    for (auto&& r : trinfo.next) {
+        leaving.erase(r);
+    }
+    if (leaving.empty()) {
+        throw std::runtime_error(format("No leaving replicas"));
+    }
+    if (leaving.size() > 1) {
+        throw std::runtime_error(format("More than one leaving replica"));
+    }
+    return *leaving.begin();
+}
+
+inet_address storage_service::host2ip(locator::host_id host) {
+    auto ip = _group0->address_map().find(raft::server_id(host.uuid()));
+    if (!ip) {
+        throw std::runtime_error(::format("Cannot map host {} to ip", host));
+    }
+    return *ip;
+}
+
+// Streams data to the pending tablet replica of a given tablet on this node.
+// The source tablet replica is determined from the current transition info of the tablet.
+future<> storage_service::stream_tablet(locator::global_tablet_id tablet) {
+    // The coordinator does not execute global token metadata barrier before jumping to "streaming" stage, so we need
+    // a barrier here to see the token metadata which is at least as recent as that of the sender.
+    auto& raft_server = _group0->group0_server();
+    co_await raft_server.read_barrier(&_abort_source);
+
```

Comment:
Shouldn't you check that the coordinator is not outdated?

Reply:
Instead of that, I check whether the request matches what's in local tablet transition info. Here's the line of thought. Instead of an RPC, we could trigger streaming from the local change of tablet_transition_info, as all the decision follows from that. There is no need for global token metadata barrier in the "streaming" stage, because it's performed in the previous stage. So it doesn't matter if the coordinator is still the same, as long as we execute what is in the current tablet_transition_info. If that migration is aborted (not impl yet) the coordinator would execute global token metadata barrier to ensure all replicas notice the abort stage and old RPCs don't run.

Reply:
> So why pass it at all? Just take it from there and check for outdated RPC the same way topology_cmd does?

You mean why pass tablet id instead of starting all sessions which target the RPC-handling node? Even though in this PR we stream all tablets in tandem, later we will wan to stream independently, so the coordinator needs to know when a given tablet finished migrating, not just all tablets on a given replica.